src/OmenThumbnailMapping.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+interface IFixedProductMarketMaker {
+    function balanceOf(address account) external view returns (uint256);
+}
+
+contract OmenThumbnailMapping {
+    // Mapping from market's address to image's IPFS hash
+    mapping(address => string) private marketAddressToIPFSHash;
+    mapping(address => address) private marketAddressToLatestImageChanger;
+
+    function get(address marketAddress) public view returns (string memory) {
+        // Get IPFS hash of thumbnail for the given market.
+        return marketAddressToIPFSHash[marketAddress];
+    }
+
+    function set(address marketAddress, string memory image_hash) public {
+        // Update IPFS hash of thumbnail for the given market.
+        requireThatSenderCanChangeImage(marketAddress);
+        marketAddressToIPFSHash[marketAddress] = image_hash;
+        marketAddressToLatestImageChanger[marketAddress] = msg.sender;
+    }
+
+    function remove(address marketAddress) public {
+        // Remove IPFS hash of thumbnail for the given market.
+        requireThatSenderCanChangeImage(marketAddress);
+        delete marketAddressToIPFSHash[marketAddress];
+        marketAddressToLatestImageChanger[marketAddress] = msg.sender;
+    }
+
+    function requireThatSenderCanChangeImage(
+        address marketAddress
+    ) public view {
+        // Verify that sender is allowed to update IPFS hash for the given market.
+        IFixedProductMarketMaker market = IFixedProductMarketMaker(
+            marketAddress
+        );
+        uint256 fundedBySender = market.balanceOf(msg.sender);
+        require(fundedBySender > 0, "Sender has no shares in the market.");
+
+        address latestImageChanger = marketAddressToLatestImageChanger[
+            marketAddress
+        ];
+        uint256 fundedByLatestImageChanger = market.balanceOf(
+            latestImageChanger
+        );
+        require(
+            latestImageChanger == address(0) ||
+                fundedBySender > 2 * fundedByLatestImageChanger,
+            "Sender don't have at least double the shares than the latest person who updated the image."
+        );
```

Comment:
Where does this requirement `2 * number_of_shares` comes from please?

Reply:
The idea is similar to challenging answers on Realitio.

In the beginning, anyone with `>0` provided liquidity in the market can change the image.

But then, if anyone wants to challenge that image, he/she needs to provide at least double the liquidity to the market of that latest image changer.

This is to prevent people from being funny and changing all the images on Omen to something funky.

Reply:
I would have suggested to use the market creator as the only entity able to update images - but since that's not trivial to obtain from Solidity, this approach is good enough for me.

Reply:
Well we have discussed together on Slack that it isn't possible. Or do you have a new idea about how it can be done? (if you are saying it's not trivial, not impossible)